### ISSUE TITLE
Make dataProvider types generic

### ISSUE THREAD

**Issue by andrico1234** [marmelab/react-admin#5894]
**Is your feature request related to a problem? Please describe.**
The types for the data provider methods aren't generic, and are often written like the following: 

```
interface UpdateParams {
  id: Identifier;
  data: any;
  previousData: Record;
}
```

This means that when I create a function like this:

```
const dataProvider: DataProvider = {
  update: (resource, params) => {
      return updateResource(data);
  },
};

async function updateResource(body: UpdateParams) {
  const { data } = body
  // fetch request
}
```

The type of body is any, even though it's likely I know what the type is at this point.

**Describe the solution you'd like**

```
interface UpdateParams<T = any> {
  id: Identifier;
  data: T;
  previousData: Record;
}
```

This should be a backward compatible type as the generic defaults to `any` which is existing behaviour.

If you're happy for this to go in, I'll be happy to make a PR to update all of the dataProvider method types.

**Comment by djhi** [comment]
Hi @andrico1234 and thanks for the suggestion! We know we have to improve the types around the dataProvider and any help is welcome! Please open a PR so that we can discuss how we do it and iterate.

**Comment by andrico1234** [comment]
Sounds good! i'll whip up something and get your thoughts

**Comment by andrico1234** [comment]
Super tiny MR [here](https://github.com/marmelab/react-admin/pull/5934)

**Comment by fzaninotto** [comment]
Fixed by #5934
